File: Experiment-1/restrain_bodies/restrain_bodies.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <cctype>
#include <locale>
#include <random>
#include <unordered_set>
#include <unordered_map>

struct Atom {
    char chain;
    int resi;
    std::string aname;
    double x, y, z;
};

// Trimming functions for whitespace
void ltrim(std::string &s) {
    s.erase(s.begin(), std::find_if(s.begin(), s.end(), [](unsigned char ch) {
        return !std::isspace(ch);
    }));
}

void rtrim(std::string &s) {
    s.erase(std::find_if(s.rbegin(), s.rend(), [](unsigned char ch) {
        return !std::isspace(ch);
    }).base(), s.end());
}

void trim(std::string &s) {
    ltrim(s);
    rtrim(s);
}

std::vector<Atom> read_structure(const std::string& pdb_filename) {
    std::ifstream pdb_file(pdb_filename);
    std::string line;
    std::vector<Atom> atoms;
    std::unordered_set<std::string> important_atoms{"CA", "P"};

    while (std::getline(pdb_file, line)) {
        if (line.substr(0, 4) == "ATOM") {
            std::string aname = line.substr(12, 4);
            trim(aname);
            if (important_atoms.find(aname) != important_atoms.end()) {
                Atom atom;
                atom.aname = aname;
                atom.chain = line[21];
                atom.resi = std::stoi(line.substr(22, 4));
                atom.x = std::stod(line.substr(30, 8));
                atom.y = std::stod(line.substr(38, 8));
                atom.z = std::stod(line.substr(46, 8));
                atoms.push_back(atom);
            }
        }
    }
    if (atoms.empty()) {
        std::cerr << "Error: No atoms found or file could not be read correctly." << std::endl;
        exit(EXIT_FAILURE);
    }
    return atoms;
}

double calc_euclidean(const Atom& i, const Atom& j) {
    return std::sqrt((j.x - i.x) * (j.x - i.x) + (j.y - i.y) * (j.y - i.y) + (j.z - i.z) * (j.z - i.z));
}

std::vector<std::pair<int, int>> get_bodies(const std::vector<Atom>& atoms, const std::unordered_map<std::string, double>& thresholds) {
    std::vector<std::pair<int, int>> bodies;
    int start = 0;

    for (size_t i = 1; i < atoms.size(); ++i) {
        if (atoms[i].chain != atoms[i - 1].chain || atoms[i].aname != atoms[i - 1].aname ||
            calc_euclidean(atoms[i - 1], atoms[i]) >= thresholds.at(atoms[i].aname)) {
            bodies.push_back(std::make_pair(start, i - 1));
            start = i;
        }
    }
    bodies.push_back(std::make_pair(start, atoms.size() - 1));  // Last body
    return bodies;
}

std::vector<std::pair<int, int>> build_restraints(const std::vector<std::pair<int, int>>& bodies, std::mt19937& generator) {
    std::vector<std::pair<int, int>> restraints;

    for (size_t i = 0; i < bodies.size(); ++i) {
        for (size_t j = i + 1; j < bodies.size(); ++j) {
            int start_i = bodies[i].first;
            int end_i = bodies[i].second;
            int start_j = bodies[j].first;
            int end_j = bodies[j].second;

            if (end_i - start_i > 0 && end_j - start_j > 0) { // Ensure there are at least two residues
                std::uniform_int_distribution<> dis_i(start_i, end_i);
                std::uniform_int_distribution<> dis_j(start_j, end_j);

                int res_i = dis_i(generator);
                int res_j = dis_j(generator);

                // Additional check to ensure different residues
                while (res_i == res_j && start_i == start_j) {
                    res_j = dis_j(generator);
                }

                restraints.push_back(std::make_pair(res_i, res_j));
            }
        }
    }
    return restraints;

}

void generate_tbl(const std::vector<Atom>& atoms, const std::vector<std::pair<int, int>>& restraints) {
    for (auto& r : restraints) {
        double dist = calc_euclidean(atoms[r.first], atoms[r.second]);
        std::cout << "assign (chain " << atoms[r.first].chain << " and resi " << atoms[r.first].resi << ")"
                  << " (chain " << atoms[r.second].chain << " and resi " << atoms[r.second].resi << ")"
                  << " " << dist << " 0.0 0.0" << std::endl;
    }
}

int main() {
    std::mt19937 generator(917);  // Seeding for reproducibility
    std::string filename = "./data/emref_1.pdb";
    std::unordered_map<std::string, double> thresholds{{"CA", 4.0}, {"P", 7.5}};

    std::vector<Atom> atom_list = read_structure(filename);
    auto bodies = get_bodies(atom_list, thresholds);
    auto restraints = build_restraints(bodies, generator);
    generate_tbl(atom_list, restraints);

    return 0;
}

File: Experiment-1/restrain_bodies/restrain_bodies.py
"""haddock3-restraints restrain_bodies subcommand.

The restrain_bodies subcommand creates distance restraints to lock several 
chains together. Useful to avoid unnatural flexibility or movement due to
sequence/numbering gaps.

Usage::
	
	haddock3-restraints restrain_bodies <structure> [--exclude] [--verbose]/
"""

# from python import sys
# from python import random
# from python import itertools
# from python import time
import sys
import random
import itertools
import time

################################################################################### 


def read_structure(pdbf):
    """
    Reads a PDB file and returns a list of parsed atoms
    """
    _atoms = {'CA', 'P'}  # Alpha-Carbon (Prot), Backbone Phosphorous (DNA)
    _altloc = {' ', 'A'}
    # print(' hello ? ')
    # print(pdbf)
    exclude = set()

    res_list = []
    with open(pdbf, 'r') as pdb_handle:
        for line in pdb_handle:
            if line.startswith('ATOM'):
                aname = line[12:16].strip()
                chain = line[21] if line[21].strip() else line[72:76].strip()
                if chain not in exclude and aname in _atoms and line[16] in _altloc:
                    resi = int(line[22:26])
                    coords = (float(line[30:38]), float(line[38:46]), float(line[46:54]))
                    res_list.append((chain, resi, aname, coords))

    if not res_list:
        # print(f'[CRITICAL] PDB File seems empty or no CA/P atoms found: {pdbf}')
        sys.exit(1)

    return res_list

def calc_euclidean(i, j):
    return ((j[0]-i[0])**2 + (j[1]-i[1])**2 + (j[2]-i[2])**2)**0.5

def get_bodies(atom_lst, prot_threshold=4.0, dna_threshold=7.5):
    bodies = []
    threshold = {'CA': prot_threshold, 'P': dna_threshold}

    body_start = 0
    for i, atom in enumerate(atom_lst[1:], start=1):
        p_atom = atom_lst[i-1]
        chain, resi, aname, xyz = atom
        p_chain, p_resi, p_aname, p_xyz = p_atom

        if (chain == p_chain) and (aname == p_aname) and calc_euclidean(xyz, p_xyz) >= threshold[aname]:
            # print(f'[DEBUG] (Internal) Body: {body_start}:{i-1}')
            bodies.append((body_start, i-1))
            body_start = i

        elif (chain != p_chain) or (aname != p_aname):
            # print(f'[DEBUG] Body: {body_start}:{i-1}')
            bodies.append((body_start, i-1))
            body_start = i

    if i is not None:  # To handle the case where the loop doesn't run
        # print(f'[DEBUG] Body: {body_start}:{i}')
        bodies.append((body_start, i))

    print(f'[INFO] Found {len(bodies)} bodies')

    return bodies


def build_restraints(bodies):
    import random
    random.seed(917)
    def pick_residues(body, max_trials=10):
        # Ensure body is a list for random.sample
        body_list = list(body)
        
        n_trials = 0
        while True:
            try:
                res_i, res_ii = random.sample(body_list, 2)
            except ValueError:
                # Likely, sample size is 1
                print('[!] One-sized body found. This may lead to problems..')
                return body_list[0], body_list[0]

            # print(f'[+++] Trial {n_trials}: {res_i} & {res_ii}')
            if abs(res_i - res_ii) > 3:
                # print(f'[++] Picked residues {res_i} & {res_ii}')
                return res_i, res_ii
            n_trials += 1
            if n_trials == max_trials:
                print(f'[!] Could not pick two unique distant residues in body after {max_trials} tries')
                return res_i, res_ii

    restraints = []

    # Use a list of tuples for bodies to avoid indexing issues
    # @par
    for i, body_i_tuple in enumerate(bodies):
        for j, body_j_tuple in enumerate(bodies):
            if i >= j:  # Avoid duplicating work
                continue

            st_body_i, en_body_i = body_i_tuple
            st_body_j, en_body_j = body_j_tuple

            res_i, res_ii = pick_residues(range(st_body_i, en_body_i + 1))
            res_j, res_jj = pick_residues(range(st_body_j, en_body_j + 1))

            # print(f'[++] Created restraint: {i}:{res_i} <--> {j}:{res_j}')
            restraints.append((res_i, res_j))
            # print(f'[++] Created restraint: {i}:{res_ii} <--> {j}:{res_jj}')
            restraints.append((res_ii, res_jj))

    return restraints


def generate_tbl(atom_lst, restraints):
    for r in restraints:
        i, j = r
        atom_i, atom_j = atom_lst[i], atom_lst[j]
        dist_ij = calc_euclidean(atom_i[3], atom_j[3])

        print(f"assign (segid {atom_i[0]} and resi {atom_i[1]} and name {atom_i[2]}) ",
              f"(segid {atom_j[0]} and resi {atom_j[1]} and name {atom_j[2]}) ",
              f"{dist_ij:.3f} 0.0 0.0")

################################################################################## 
# Set random seed to have reproducibility
# random.seed(917)

def restrain_bodies(structure):

    atom_lst = read_structure(structure)
    # print('atoms got:')
    # print(atom_lst)
    bodies = get_bodies(atom_lst)
    # print("got bodies:")
    # print(bodies)
    restraints = build_restraints(bodies)
    # print("got restrains:")
    # print(restraints)
    generate_tbl(atom_lst, restraints)

################################################################################## test: 
if __name__ == "__main__":
    start_time = time.time()
    restrain_bodies('./data/emref_1.pdb')
    duration = time.time() - start_time
    print(duration)